homework/81955/hw1/Lexer.cpp

```diff
@@ -0,0 +1,526 @@
+    keywordTable["go"] = TokenType::Go;
+    keywordTable["goto"] = TokenType::GoTo;
+    keywordTable["if"] = TokenType::If;
+    keywordTable["import"] = TokenType::Import;
+    keywordTable["Interface"] = TokenType::Interface;
+    keywordTable["map"] = TokenType::Map;
+    keywordTable["package"] = TokenType::Package;
+    keywordTable["range"] = TokenType::Range;
+    keywordTable["return"] = TokenType::Return;
+    keywordTable["select"] = TokenType::Select;
+    keywordTable["struct"] = TokenType::Struct;
+    keywordTable["switch"] = TokenType::Switch;
+    keywordTable["type"] = TokenType::Type;
+    keywordTable["var"] = TokenType::Var;
+    keywordTable["nil"] = TokenType::Nil;
+    keywordTable["true"] = TokenType::True;
+    keywordTable["false"] = TokenType::False;
+    keywordTable["int"] = TokenType::Int;
+    keywordTable["string"] = TokenType::String;
+    keywordTable["float"] = TokenType::Float;
+    keywordTable["bool"] = TokenType::Bool;
+}
+inline bool Tokenizer::match(char c)
+{
+    if (code[index] == c)
+    {
+        nextSymbol();
+        return true;
+    }
+    return false;
+}
+
+inline void Tokenizer::nextSymbol()
+{
+    index++;
+}
+
+inline void Tokenizer::previousSymbol()
+{
+    index--;
+}
+
+std::string Tokenizer::parseWhitespace()
+{
+    std::string whitespace;
+    while (code[index] == ' ' || code[index] == '\n' || code[index] == '\t')
+    {
+        whitespace += code[index];
+        nextSymbol();
+    }
+    if (whitespace != "")
+        state = true;
+    else
+        state = false;
+    return whitespace;
+}
+
+std::string Tokenizer::parseComment()
+{
+    if (!match('/'))
+    {
+        state = false;
+        return "";
+    }
+
+    std::string comment = "/";
+
+    if (match('/'))
+    {
+        comment += '/';
+        while (!Helper::isEof(code[index]) && !Helper::isNewLine(code[index]))
+        {
+            comment += code[index];
+            nextSymbol();
+        }
+        state = true;
+        return comment;
+    }
+
+    else if (match('*'))
+    {
+        comment += '*';
+        while (code[index] != '*' && code[index + 1] != '/' && !Helper::isEof(code[index]))
+        {
+            comment += code[index];
+            nextSymbol();
+        }
+        if (!Helper::isEof(code[index]))
+        {
+            nextSymbol();
+            nextSymbol();
+            comment += "*/";
+        }
+        state = true;
+        return comment;
+    }
+    else
+        previousSymbol();
+
+    state = false;
+    return "";
+}
+
+std::string Tokenizer::parseString()
+{
+
+    if (!Helper::isStringLiteral(code[index]))
+    {
+        state = false;
+        return "";
+    }
+
+    char c = code[index];
+    std::string res;
+    res += c;
+    nextSymbol();
+    while (!match(c) && !Helper::isEof(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+    }
+    if (code[index - 1] == c)
+    {
+        res += c;
+    }
+    state = true;
+    return res;
```

Comment:
In all the methods, instead of constantly accumulating characters in the resulting string, you can keep the `index` at the start of the function - say `startIndex`, and return the `[startIndex, currentIndex)` substring of the input.